A94-subchannel-otel-metrics.md

```diff
@@ -0,0 +1,128 @@
+-------------------- | -----------
+GOAWAY <ERROR_CODE>  | HTTP2 GOAWAY frame with error code for example (“GOAWAY NO_ERROR”, “GOAWAY PROTOCOL_ERROR”, “GOAWAY ENHANCE_YOUR_CALM”). The list of error codes is available in [RFC 9113](https://www.rfc-editor.org/rfc/rfc9113.html#name-error-codes).
+subchannel shutdown  | The subchannel was shutdown. This can happen due to reasons such as the parent channel shutting down, channel becoming idle, the load balancing policy changing due to a resolver update or a change in list of endpoint addresses.
+connection reset     | Connection was reset (eg. ECONNRESET, WSAECONNERESET)
+connection timed out | Connection timed out, also includes connections closed due to gRPC keepalives.
+connection aborted   | Connection was aborted
+socket error         | Any socket error not covered by “connection reset”, “connection timed out” and “connection aborted”. Implementations that are not able to differentiate between the different socket error codes should also use this.
+unknown              | Catch-all for all other reasons.
+
+For a given connection, there can be multiple reasons reported to the subchannel
+for disconnection. For example, a connection could have seen a GOAWAY frame with
+`ENHANCE_YOUR_CALM` and then a socket error Broken Pipe. In such cases, the
+first seen reason should be chosen, `GOAWAY ENHANCE_YOUR_CALM` in this case.
+
+We might add more error cases to this in the future.
+
+## Rationale
+
+### Renaming pick-first metrics
+
+The existing pick-first metrics provides stats on subchannel disconnections and
+connection attempts as viewed from the perspective of the pick-first lb policy.
+[A61: IPv4 and IPv6 Dualstack Backend Support](A61-IPv4-IPv6-dualstack-backends.md)
+made pick-first lb policy the universal leaf policy. For users unfamiliar with
+this, it will come as a surprise when metrics for pick-first lb policy are
+populated when round_robin lb policy is configured (for example). Additionally,
+the pick-first metrics are defined from the perspective of the channel. This
+means that if subchannels are shared between multiple channels (as is the case
+for gRPC Core and its wrapped languages - C++, Python), we will double-count the
+disconnections/connection attempts.
+
+Renaming/moving the pick-first metrics to subchannel makes this more intuitive,
+and fixes the double-counting problem.
+
+### Metric for open connections
+
+Moving the metrics down to subchannel potentially allows us to calculate the
+number of open connections by subtracting `grpc.subchannel.disconnections` from
+`grpc.subchannel.connection_attempts_succeeded`. This method does not work for
+exporters recording counters per period in a way that does not allow for a
+simple subtraction of the two counters
+(https://github.com/grpc/grpc/issues/34886).
+
+Adding an explicit metric that records the number of open connections avoids
+this.
+
+### Use “connection timed out” as disconnect error for gRPC Keepalive timeouts
```

Comment:
This section is a little unclear.  I think the title should be something like "Combining connection timeouts and keepalives into a single disconnection error", and we should explain that since we generally use the same timeouts for both keepalives and TCP_USER_TIMEOUT, it will essentially be a race as to which one fires first, so there's not really any point in splitting these into two different buckets.

Reply:
amended